src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs

```diff
@@ -0,0 +1,474 @@
+            {
+                _diagnosticIds = null;
+            }
+
+            public IDELatestDiagnosticGetter(DiagnosticIncrementalAnalyzer owner, ImmutableHashSet<string> diagnosticIds, Solution solution, ProjectId projectId, bool includeSuppressedDiagnostics) :
+                this(owner, diagnosticIds, solution, projectId, documentId: null, includeSuppressedDiagnostics: includeSuppressedDiagnostics)
+            {
+            }
+
+            public IDELatestDiagnosticGetter(DiagnosticIncrementalAnalyzer owner, ImmutableHashSet<string> diagnosticIds, Solution solution, ProjectId projectId, DocumentId documentId, bool includeSuppressedDiagnostics) :
+                base(owner, solution, projectId, documentId, id: null, includeSuppressedDiagnostics: includeSuppressedDiagnostics)
+            {
+                _diagnosticIds = diagnosticIds;
+            }
+
+            public async Task<ImmutableArray<DiagnosticData>> GetProjectDiagnosticsAsync(CancellationToken cancellationToken)
+            {
+                if (CurrentSolution == null)
+                {
+                    return GetDiagnosticData();
+                }
+
+                DocumentId nullTargetDocumentId = null;
+
+                if (CurrentProjectId != null)
+                {
+                    await AppendDiagnosticsAsync(CurrentProject, nullTargetDocumentId, SpecializedCollections.EmptyEnumerable<DocumentId>(), cancellationToken).ConfigureAwait(false);
+                    return GetDiagnosticData();
+                }
+
+                await AppendDiagnosticsAsync(CurrentSolution, cancellationToken).ConfigureAwait(false);
+                return GetDiagnosticData();
+            }
+
+            protected override bool ShouldIncludeDiagnostic(DiagnosticData diagnostic)
+            {
+                return _diagnosticIds == null || _diagnosticIds.Contains(diagnostic.Id);
+            }
+
+            protected override async Task AppendDiagnosticsAsync(Project project, DocumentId targetDocumentId, IEnumerable<DocumentId> documentIds, CancellationToken cancellationToken)
+            {
+                // when we return cached result, make sure we at least return something that exist in current solution
+                if (project == null)
+                {
+                    return;
+                }
+
+                // get analyzers that are not suppressed.
+                // REVIEW: IsAnalyzerSuppressed call seems can be quite expensive in certain condition. is there any other way to do this?
+                var stateSets = StateManager.GetOrCreateStateSets(project).Where(s => ShouldIncludeStateSet(project, s)).ToImmutableArrayOrEmpty();
+
+                var analyzerDriverOpt = await Owner._compilationManager.CreateAnalyzerDriverAsync(project, stateSets, IncludeSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
```

Comment:
@heejaechang Just an FYI: We probably may have to create a map in this type (just like http://index/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs,398) to ensure that we re-use the instance of CompilationWithAnalyzers when computing diagnostics for different documents in the same project, if any GetDiagnosticsAsync code path ends up invoking this method in parallel for documents. I see that v1 used to do so [here](http://index/#Microsoft.CodeAnalysis.Features/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_GetDiagnostics.cs,209), but none of the code you added is doing so, hence not having a map should probably be fine for now. Just something to keep in mind for you.